Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp

```diff
@@ -440,6 +440,10 @@ void PaymentRequest::show(Document& document, RefPtr<DOMPromise>&& detailsPromis
 
 void PaymentRequest::abortWithException(Exception&& exception)
 {
+    // If state is "closed", then the request has already been aborted.
+    if (m_state == State::Closed)
+        return;
```

Comment:
We have code that calls an internal function, which may return an exception, it is returns an exception, we call `abortWithException()` with that exception. The issue is that the internal function may return an exception because we've already aborted. In this case, we would end up calling `abortWithException()` on something that is already aborted.

At first, I had the `m_state != State::Closed` at the couple of call sites that call abortWithException() with a "forwarded" exception but then I decided to simplify the code and move the check inside abortWithException().

An example of such call site is:
```
        auto exception = activePaymentHandler->merchantValidationCompleted(m_merchantSessionPromise->result());
        if (exception.hasException()) {
            abortWithException(exception.releaseException());
            return;
        }
```

Which we would have updated to:
```
        auto exception = activePaymentHandler->merchantValidationCompleted(m_merchantSessionPromise->result());
        if (exception.hasException()) {
            if (m_state != State::Closed)
                abortWithException(exception.releaseException());
            return;
        }
```

Please let me know what you think now that I've clarified.

Reply:
i see.  i would almost prefer the latter so that we could then know exactly where this kind of double abort can happen and handle cases where it's unexpected, but i also see the benefit of having it be generalized inside `PaymentRequest::abortWithException`